File: apps/mobile/android/app/src/main/java/xyz/speakeasyapp/app/voicefilter/dsp/GranularPitchShifter.kt
package xyz.speakeasyapp.app.voicefilter.dsp

/**
 * Tape-head granular pitch shifter — the simplest "voice mask" DSP.
 *
 * Maintains a ring buffer of recent input samples; a read head advances
 * at `factor` samples per output sample (factor > 1 ⇒ pitch up,
 * factor < 1 ⇒ pitch down). When the read head approaches the write
 * head (from either side, since they share a ring), it cross-fades
 * with the sample read [grainSize] samples earlier — the classic
 * granular trick that keeps the "tape" scrubbing continuously even
 * though the rates differ.
 *
 * This shifts BOTH pitch and formants by the same factor; that's the
 * locked v1 behavior. A future PR can swap this for an independent
 * pitch+formant shifter (phase vocoder + cepstral whitening) behind
 * the same [SampleFilter][xyz.speakeasyapp.app.voicefilter.SampleFilter]
 * interface without touching the WebRTC ADM fork.
 *
 * Allocation-free per-frame: all buffers sized in the constructor.
 * Not thread-safe; the single WebRTC audio-record thread is the only
 * intended caller.
 */
internal class GranularPitchShifter(
    /**
     * Length of one grain in samples. ~42ms at 48kHz is a good
     * compromise between artifact frequency (smaller grain = louder
     * crackle) and pitch resolution (larger grain = audible delay).
     */
    private val grainSize: Int = 2048,
    /**
     * Cross-fade width in samples. Wider = smoother but more
     * "smear"; narrower = sharper but more crackle. 256 ≈ 5ms at
     * 48kHz, a small fraction of the grain.
     */
    private val crossFade: Int = 256,
) {
  init {
    require(grainSize > 0 && grainSize and (grainSize - 1) == 0) {
      "grainSize must be a positive power of two for the ring-wrap arithmetic"
    }
    require(crossFade in 1..(grainSize / 4)) {
      "crossFade must be in (0, grainSize/4]"
    }
  }

  /** Ring buffer holding the most recent inputs in float-normalized form. */
  private val ringSize = grainSize * 2
  private val ring = FloatArray(ringSize)
  private var writeIdx = 0

  /**
   * Tape read head as a `Double` so fractional rates accumulate without
   * drift. Wraps into `[0, ringSize)` each frame.
   *
   * Starts diametrically opposite the write head so the cross-fade
   * doesn't fire on frame 0: at factor=1.0 the heads keep their
   * `ringSize/2` separation and the output is a clean passthrough; at
   * shift factors close to 1, the first cross-fade lap is pushed to
   * the ~0.4–0.8s mark (well past the test/UX warmup).
   */
  private var readPos = (grainSize).toDouble()

  /**
   * Filter one frame of `n` PCM16 samples in-place — reads from
   * [input] starting at [inOffset], writes to [output] starting at
   * [outOffset]. Both buffers MAY be the same array; the ring keeps
   * its own copy.
   *
   * @param factor positive pitch-shift factor. `1.0` is a passthrough.
   */
  fun process(
      input: ShortArray,
      inOffset: Int,
      output: ShortArray,
      outOffset: Int,
      n: Int,
      factor: Float,
  ) {
    // 1) Snapshot input into the ring buffer first so the read head
    // never crosses fresh data this frame (avoids self-feedback when
    // input === output).
    var w = writeIdx
    for (i in 0 until n) {
      ring[w] = input[inOffset + i] / 32768f
      w = (w + 1) and (ringSize - 1)
    }
    writeIdx = w

    // 2) Render output via the tape head.
    val crossFadeF = crossFade.toFloat()
    val ringMask = ringSize - 1
    val grainSizeI = grainSize
    var pos = readPos
    val factorD = factor.toDouble()
    for (i in 0 until n) {
      val posI = pos.toInt()
      val posF = (pos - posI).toFloat()
      val a0 = posI and ringMask
      val a1 = (posI + 1) and ringMask
      val sa = ring[a0] + (ring[a1] - ring[a0]) * posF

      // The grain-back position lags by one grainSize, so wrapping
      // there gives us a sample whose cross-fade window doesn't
      // collide with the write head at the same instant.
      val bI = posI - grainSizeI
      val b0 = bI and ringMask
      val b1 = (bI + 1) and ringMask
      val sb = ring[b0] + (ring[b1] - ring[b0]) * posF

      // Forward distance to write head, normalized into [0, ringSize).
      val dist = (writeIdx - posI - 1 + ringSize) and ringMask

      val mix =
          when {
            dist < crossFade -> dist.toFloat() / crossFadeF
            dist >= ringSize - crossFade -> (ringSize - dist).toFloat() / crossFadeF
            else -> 1f
          }
      val mixed = sa * mix + sb * (1f - mix)

      // Clamp to PCM16 range.
      val clamped =
          if (mixed > 1f) 1f else if (mixed < -1f) -1f else mixed
      output[outOffset + i] = (clamped * 32767f).toInt().toShort()

      pos += factorD
      if (pos >= ringSize) pos -= ringSize
      if (pos < 0) pos += ringSize
    }
    readPos = pos
  }

  fun reset() {
    ring.fill(0f)
    writeIdx = 0
    readPos = grainSize.toDouble()
  }
}

File: apps/mobile/ios/SpeakeasyBridges/VoiceFilter/dsp/GranularPitchShifter.swift
//
//  GranularPitchShifter.swift
//  Speakeasy — Phase 5j Private Call voice filter
//
//  Swift port of
//  apps/mobile/android/.../voicefilter/dsp/GranularPitchShifter.kt
//
//  Tape-head granular pitch shifter — the simplest "voice mask"
//  DSP. Both pitch and formants move by the same factor; that's
//  the locked v1 behavior. A future PR can swap this for an
//  independent pitch + formant shifter (phase vocoder + cepstral
//  whitening) behind the same SampleFilter interface without
//  touching the RTCAudioDevice plumbing.
//
//  Allocation-free per-frame: all buffers sized in init().
//  Not thread-safe; the single AVAudioEngine capture thread is the
//  only intended caller.
//

import Foundation

final class GranularPitchShifter {

  /// `grainSize` must be a positive power of two so the ring-wrap
  /// arithmetic stays a cheap mask. 1024 ≈ 21ms at 48kHz — halved
  /// from the previous 2048 (rc.16 dogfood: filter delay was
  /// pushing end-to-end call latency past the "feels natural"
  /// threshold). Trade-off accepted: more audible granular
  /// crackle at this grain size, especially on sustained vowels.
  /// Phase 2 swaps the algorithm for a phase vocoder which fixes
  /// both — until then, latency is the more painful axis.
  private let grainSize: Int
  /// Cross-fade width in samples. 256 ≈ 5ms at 48kHz.
  private let crossFade: Int

  private let ringSize: Int
  private let ringMask: Int
  private var ring: [Float]

  private var writeIdx: Int = 0
  /// Read head as Double so fractional rates don't drift. Wraps
  /// into [0, ringSize) each frame. Starts diametrically opposite
  /// writeIdx so factor=1 is a clean passthrough (heads stay
  /// `grainSize` apart) and shift factors near 1 push the first
  /// cross-fade lap to ~0.4–0.8s — past the warmup window.
  private var readPos: Double

  init(grainSize: Int = 1024, crossFade: Int = 256) {
    precondition(grainSize > 0 && grainSize & (grainSize - 1) == 0,
                 "grainSize must be a positive power of two")
    precondition(crossFade > 0 && crossFade <= grainSize / 4,
                 "crossFade must be in (0, grainSize/4]")
    self.grainSize = grainSize
    self.crossFade = crossFade
    self.ringSize = grainSize * 2
    self.ringMask = self.ringSize - 1
    self.ring = [Float](repeating: 0, count: ringSize)
    self.readPos = Double(grainSize)
  }

  /// Process `n` mono PCM16 samples through the shifter. `input`
  /// and `output` MAY alias; the ring keeps its own copy.
  func process(
    input: UnsafePointer<Int16>,
    output: UnsafeMutablePointer<Int16>,
    count n: Int,
    factor: Float
  ) {
    // 1) Snapshot input into the ring so the read head never crosses
    // fresh data this frame (avoids self-feedback when input == output).
    var w = writeIdx
    for i in 0..<n {
      ring[w] = Float(input[i]) / 32768.0
      w = (w + 1) & ringMask
    }
    writeIdx = w

    // 2) Render output via the tape head.
    let crossFadeF = Float(crossFade)
    var pos = readPos
    let factorD = Double(factor)
    for i in 0..<n {
      let posI = Int(pos)
      let posF = Float(pos - Double(posI))
      let a0 = posI & ringMask
      let a1 = (posI &+ 1) & ringMask
      let sa = ring[a0] + (ring[a1] - ring[a0]) * posF

      // Grain-back position lags by one grainSize — diametrically
      // opposite read when grainSize = ringSize/2, which it is.
      let bI = posI &- grainSize
      let b0 = bI & ringMask
      let b1 = (bI &+ 1) & ringMask
      let sb = ring[b0] + (ring[b1] - ring[b0]) * posF

      // Forward distance to write head in [0, ringSize).
      let dist = (writeIdx &- posI &- 1 &+ ringSize) & ringMask

      let mix: Float
      if dist < crossFade {
        mix = Float(dist) / crossFadeF
      } else if dist >= ringSize - crossFade {
        mix = Float(ringSize - dist) / crossFadeF
      } else {
        mix = 1.0
      }
      let mixed = sa * mix + sb * (1.0 - mix)

      let clamped = max(-1.0, min(1.0, mixed))
      output[i] = Int16(clamped * 32767.0)

      pos += factorD
      if pos >= Double(ringSize) { pos -= Double(ringSize) }
      if pos < 0 { pos += Double(ringSize) }
    }
    readPos = pos
  }

  func reset() {
    for i in 0..<ringSize { ring[i] = 0 }
    writeIdx = 0
    readPos = Double(grainSize)
  }
}
